### PR TITLE
Show Thumbnails in Dataset Table

### LINKED ISSUE #7481: Parsing dataset tags with special characters is broken again
Inserting tags with commas leads to them being reported as two tags, with added quotes in them. Parsing the array literal seems to be broken

### LINKED ISSUE #7482: Show thumbnails in dataset list
With the recent improvements for thumbnails, they could now be shown always in the dataset list.

### PR DESCRIPTION
**Pull request by fm3**
### Description

- Introduces new design for the dashboard dataset list:
  - thumbnails replace the "type" column
  - folders and inactive datasets have icons in place of the thumbnails
  - tags are now rendered beneath the dataset name, no longer in a separate column
- Fixes a backend bug where tags with special characters (`" \ ,`) would get unexpectedly split

### Screenshots
image

### URL of deployed dev instance (used for testing):
- https://datasettablethumbnails.webknossos.xyz

### Steps to test:
- Open dashboard dataset list, browse around, should look pretty
- Add and remove tags with and without special characters

### TODOs:
- [x] decide how to handle multi-layer datasets (blending? apply mapping? or select a layer?)
- [x] backend
  - [x] include `layerNames` list in `APIDatasetCompact`
  - [x] avoid duplicated layer names (e.g. connectome_file_test_dataset) (double left join…)
  - [x] fix parsing sql array literals
  - [x] doublecheck that updating view config clears thumbnail cache for that dataset
- [x] frontend
  - [x] unify tag + name column
  - [x] add thumbnail image
  - [x] user layerNames for thumbnail layer selection
  - [x] styling
    - [x] bigger font? name is more important than tags
    - [x] tweak margins?
    - [x] styling for folders
    - [x] placeholder image for zero-layer datasets
  - [x] remove type column, have folder icon be same size as thumbnails
  - [x] repair dataset to dataset compact function
  - [x] error when entering search query (clearing localstorage helps)
  - [x] have someone who knows how to write frontend code look at it
  - [x] if no tags, add "Add Tag" label to the plus icon
  - [x] less aggressive colors for the tags
  - [x] new placeholder icons

### Follow-Up
 - #7487
 - #7490

### Issues:
- fixes #7481
- fixes #7482

------
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)

### CODE REVIEW

File: test/backend/SqlEscapingTestSuite.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.scalatestplus.play.PlaySpec
+import utils.sql.{SqlEscaping, SqlTypeImplicits}
+
+class SqlEscapingTestSuite extends PlaySpec with SqlTypeImplicits with SqlEscaping {
+  "SQL Escaping" should {
+    "not change plain string" in {
+      assert(escapeLiteral("hello") == "'hello'")
+    }
+    "escape string with 's" in {
```
Comment (frcroth) [inline test/backend/SqlEscapingTestSuite.scala:11]:
```suggestion
    "escape string with apostrophes" in {-
```

Comment (frcroth) [reply test/backend/SqlEscapingTestSuite.scala:11]: I think that is easier to read

File: frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx [hunk 2]
```diff
@@ -278,24 +278,45 @@ class DatasetRenderer {
...
   renderNameColumn() {
+    let selectedLayerName = null;
+    if (this.data.colorLayerNames.length > 0) {
+      selectedLayerName = this.data.colorLayerNames[0];
+    } else if (this.data.segmentationLayerNames.length > 0) {
+      selectedLayerName = this.data.segmentationLayerNames[0];
+    }
+    const imgSrc = selectedLayerName
+      ? `/api/datasets/${this.data.owningOrganization}/${this.data.name}/layers/${selectedLayerName}/thumbnail?w=200&h=200`
+      : "/assets/images/inactive-dataset-thumbnail.svg";
```
[the remark this answers is not in the document]
Comment (fm3) [reply frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx:289]: Good point, I wrote an issue tracking this https://github.com/scalableminds/webknossos/issues/7490

**Comment by normanrz** [comment]
Sweet!

File: frontend/javascripts/types/api_flow_types.ts [hunk 3]
```diff
@@ -199,12 +199,26 @@ export type APIDatasetCompactWithoutStatus = Pick<
...
+  let segmentationLayerNames = [];
+  if (dataset.dataSource) {
+    for (const layer of dataset.dataSource.dataLayers) {
+      if (layer.category === "segmentation") {
+        segmentationLayerNames.push(layer.name);
+      } else if (layer.category === "color") {
+        colorLayerNames.push(layer.name);
+      }
+    }
+  }
```
Comment (philippotto) [inline frontend/javascripts/types/api_flow_types.ts:220]:
```suggestion
  const [colorLayerNames, segmentationLayerNames] = _.partition(
    dataset.dataSource.dataLayers,
    (layer) => layer.category === "segmentation",
  ).map((layers) => layers.map((layer) => layer.name).sort());
```

not sure why you checked `dataset.dataSource`? For an `APIDataset`, the datasource should always exist (unlike `APIMaybeUnimportedDataset`).

File: frontend/javascripts/types/api_flow_types.ts [hunk 4]
```diff
@@ -217,6 +231,8 @@ export function convertDatasetToCompact(dataset: APIDataset): APIDatasetCompact
     status: dataset.dataSource.status,
     tags: dataset.tags,
     isUnreported: dataset.isUnreported,
+    colorLayerNames: colorLayerNames.sort(),
+    segmentationLayerNames: segmentationLayerNames.sort(),
```
Comment (philippotto) [inline frontend/javascripts/types/api_flow_types.ts:235]:
```suggestion
    colorLayerNames: colorLayerNames,
    segmentationLayerNames: segmentationLayerNames,
```

I integrated this in my previous comment suggestion.

File: frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx [hunk 5]
```diff
@@ -278,24 +278,45 @@ class DatasetRenderer {
     return <FileOutlined style={{ fontSize: "18px" }} />;
   }
   renderNameColumn() {
+    let selectedLayerName = null;
+    if (this.data.colorLayerNames.length > 0) {
+      selectedLayerName = this.data.colorLayerNames[0];
+    } else if (this.data.segmentationLayerNames.length > 0) {
+      selectedLayerName = this.data.segmentationLayerNames[0];
+    }
```
Comment (philippotto) [inline frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx:286]:
```suggestion
    const selectedLayerName: string | null = this.data.colorLayerNames[0] || this.data.segmentationLayerNames[0];
```

File: frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx [hunk 6]
```diff
@@ -278,24 +278,45 @@ class DatasetRenderer {
...
+    const imgSrc = selectedLayerName
+      ? `/api/datasets/${this.data.owningOrganization}/${this.data.name}/layers/${selectedLayerName}/thumbnail?w=200&h=200`
```
Comment (philippotto) [inline frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx:288]: is it intentional that the thumbnails are requested in 200x200, but there are only rendered in 80x80? is this good for high dpi screens?

**Review by philippotto — COMMENTED** [review]
very cool stuff :100: only left smaller feedback.

File: frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx (hunk 6 above)
Comment (fm3) [reply frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx:283]: Almost intentional 😅 I had it 100×100 in CSS first, with 200×200 graphics, allowing for screens with px-to-device-pixel factor of 2, which is the most common currently, other than 1.
As I changed the css size to 80×80, I’ll adapt this to use 160×160 then :)

Comment (philippotto) [reply frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx:283]: How about defining a module constant which is then used in the url as well as in an inline `style`? :) something like this:

```
const THUMBNAIL_EXTENT = 100;

`...thumbnail?w=${THUMBNAIL_EXTENT}&h=${THUMBNAIL_EXTENT}`
...<img style={{width: 2 * THUMBNAIL_EXTENT}}...
```

Comment (philippotto) [reply frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx:283]: in the top level scope below the imports is fine :)

**Review by philippotto — APPROVED** [review]
Great stuff :+1: I'm wondering whether the thumbnails will be a noticable performance hit. The first page of the dataset table can contain up to 50 datasets. This will incur 50 additional requests (can be cached of course). One could think about deferring the loading of a thumbnail (e.g., by a simple delay or by waiting until it is scrolled into view).
However, on the dev instance it's plenty fast. Let's merge it and see how it behaves in production.

**Pull request merged**